File: 1930/solution.ts
function countPalindromicSubsequence(s: string): number {
  const left = new Map<string, number>();
  const right = new Map<string, number>();
  let res = 0;
  for (let i = 0; i < s.length; i++) {
    if (left.has(s[i])) {
      right.set(s[i], i);
    } else {
      left.set(s[i], i);
    }
  }
  for (const key of right.keys()) {
    const chars = new Set<string>();
    for (let i = left.get(key)! + 1; i < right.get(key)!; i++) {
      chars.add(s[i]);
    }
    res += chars.size;
  }
  return res;
}

File: 1930/solution.go
package main

func countPalindromicSubsequence(s string) int {
	res := make(map[string]struct{})
	left := make(map[byte]struct{})
	right := make(map[byte]int)

	for i := 0; i < len(s); i++ {
		right[s[i]]++
	}

	for i := 0; i < len(s); i++ {
		char := s[i]
		right[char]--
		if right[char] == 0 {
			delete(right, char)
		}

		for leftChar := range left {
			if _, ok := right[leftChar]; ok {
				palindrome := string(leftChar) + string(char) + string(leftChar)
				res[palindrome] = struct{}{}
			}
		}
		left[char] = struct{}{}
	}
	return len(res)
}

// time O(n)
// space O(1)
